[Open in Colab](https://colab.research.google.com/github/WoolierBrooks/DementiaDetect-Speech/blob/main/scripts/adress_2020/00_Complete_nb.ipynb)

Importar bibliotecas

In [1]:
import pandas as pd
import numpy as np

Importar o dataset de treino

In [2]:
dataset = pd.read_csv("adress_20_db_compare_train.csv")
X_train = dataset.iloc[:, :-1].values
y_train = dataset.iloc[:, -1].values
print(X_train, X_train.shape)
print(y_train, y_train.shape)

[[1.97e+00 8.11e-02 2.04e-01 ... 8.05e+01 1.34e+02 8.04e+01]
 [3.37e+00 1.42e-02 0.00e+00 ... 7.24e+01 1.12e+02 7.60e+01]
 [2.83e+00 7.86e-01 9.65e-01 ... 7.27e+01 1.22e+02 7.67e+01]
 ...
 [3.53e+00 2.31e-01 0.00e+00 ... 9.54e+01 1.55e+02 9.05e+01]
 [3.37e+00 4.64e-01 0.00e+00 ... 9.83e+01 1.59e+02 8.86e+01]
 [3.63e+00 6.01e-01 6.24e-01 ... 8.99e+01 1.64e+02 9.13e+01]] (108, 6373)
[0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.
 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.
 0. 0. 0. 0. 0. 0. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1.
 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1.
 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1.] (108,)


Importando o dataset de teste

In [3]:
dataset = pd.read_csv("adress_20_db_compare_test.csv")
X_test = dataset.iloc[:, :-1].values
y_test = dataset.iloc[:, -1].values
print(X_test, X_test.shape)
print(y_test, y_test.shape)

[[3.42e+00 6.32e-01 8.09e-01 ... 8.16e+01 1.19e+02 7.26e+01]
 [5.00e-01 4.22e-01 0.00e+00 ... 8.08e+01 1.61e+02 7.93e+01]
 [2.50e+00 7.82e-01 5.27e-04 ... 6.96e+01 1.32e+02 7.47e+01]
 ...
 [3.33e+00 2.27e-01 8.64e-01 ... 8.02e+01 1.34e+02 8.00e+01]
 [3.61e+00 6.83e-01 9.75e-01 ... 8.05e+01 1.26e+02 8.12e+01]
 [1.61e+00 5.94e-01 1.60e-01 ... 7.71e+01 1.45e+02 8.42e+01]] (48, 6373)
[0. 0. 1. 0. 1. 1. 0. 1. 1. 1. 0. 1. 0. 1. 0. 0. 1. 0. 0. 1. 0. 1. 1. 0.
 0. 1. 0. 1. 1. 1. 1. 1. 1. 0. 1. 0. 0. 0. 1. 0. 1. 0. 0. 1. 0. 1. 0. 0.] (48,)


Completando dados faltantes

In [4]:
from sklearn.impute import SimpleImputer

# Conjunto de treino
imputer = SimpleImputer(missing_values=np.nan, strategy="mean")
imputer.fit(X_train)
X_train = imputer.transform(X_train)
print(X_train, X_train.shape)

[[1.97e+00 8.11e-02 2.04e-01 ... 8.05e+01 1.34e+02 8.04e+01]
 [3.37e+00 1.42e-02 0.00e+00 ... 7.24e+01 1.12e+02 7.60e+01]
 [2.83e+00 7.86e-01 9.65e-01 ... 7.27e+01 1.22e+02 7.67e+01]
 ...
 [3.53e+00 2.31e-01 0.00e+00 ... 9.54e+01 1.55e+02 9.05e+01]
 [3.37e+00 4.64e-01 0.00e+00 ... 9.83e+01 1.59e+02 8.86e+01]
 [3.63e+00 6.01e-01 6.24e-01 ... 8.99e+01 1.64e+02 9.13e+01]] (108, 6373)


In [5]:
# Conjunto de teste
imputer = SimpleImputer(missing_values=np.nan, strategy="mean")
imputer.fit(X_test)
X_test = imputer.transform(X_test)
print(X_test, X_test.shape)

[[3.42e+00 6.32e-01 8.09e-01 ... 8.16e+01 1.19e+02 7.26e+01]
 [5.00e-01 4.22e-01 0.00e+00 ... 8.08e+01 1.61e+02 7.93e+01]
 [2.50e+00 7.82e-01 5.27e-04 ... 6.96e+01 1.32e+02 7.47e+01]
 ...
 [3.33e+00 2.27e-01 8.64e-01 ... 8.02e+01 1.34e+02 8.00e+01]
 [3.61e+00 6.83e-01 9.75e-01 ... 8.05e+01 1.26e+02 8.12e+01]
 [1.61e+00 5.94e-01 1.60e-01 ... 7.71e+01 1.45e+02 8.42e+01]] (48, 6373)


Escalonamento de características

In [6]:
from sklearn.preprocessing import StandardScaler

sc = StandardScaler()
X_train = sc.fit_transform(X_train)
X_test = sc.transform(X_test)
print(X_train, X_train.shape)
print(X_test, X_test.shape)

[[-0.55697137 -1.17020103 -0.30053333 ...  0.5518206   0.40652078
   0.39635631]
 [ 0.29049993 -1.3551359  -0.88649157 ... -0.04107139 -0.66812694
   0.08181303]
 [-0.03638185  0.77838779  1.88532069 ... -0.01911243 -0.1796507
   0.131854  ]
 ...
 [ 0.3873538  -0.75582528 -0.88649157 ...  1.64244907  1.43232087
   1.11837612]
 [ 0.29049993 -0.11173223 -0.88649157 ...  1.85471904  1.62771137
   0.98255061]
 [ 0.44788746  0.26698343  0.90585128 ...  1.23986809  1.87194949
   1.17556581]] (108, 6373)
[[ 0.32076677  0.35267821  1.43723498 ...  0.63233679 -0.32619358
  -0.16124315]
 [-1.44681623 -0.22783484 -0.88649157 ...  0.57377956  1.72540662
   0.31772049]
 [-0.23614295  0.76733039 -0.88497785 ... -0.24602171  0.30882553
  -0.01112022]
 ...
 [ 0.26628647 -0.76688267  1.59521391 ...  0.52986163  0.40652078
   0.36776146]
 [ 0.43578073  0.49365996  1.91404413 ...  0.5518206   0.01573979
   0.453546  ]
 [-0.77489256  0.247633   -0.42691648 ...  0.30295235  0.94384464
   0.66800733]] (48, 

Seleção univariada de características

In [7]:
from sklearn.feature_selection import SelectKBest
from sklearn.feature_selection import f_classif

print(X_train.shape)
print(X_test.shape)

# Inicializando e ajustando o seletor de características no conjunto de treino
selector = SelectKBest(f_classif, k=637)
X_train_new = selector.fit_transform(X_train, y_train)

# Aplicando as mesmas características selecionadas para o conjunto de teste
X_test_new = selector.transform(X_test)

print(X_train_new.shape)
print(X_test_new.shape)

(108, 6373)


(48, 6373)
(108, 637)
(48, 637)


C:\Users\Lenovo\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.10_qbz5n2kfra8p0\LocalCache\local-packages\Python310\site-packages\sklearn\feature_selection\_univariate_selection.py:112: UserWarning: Features [2825] are constant.
  warnings.warn("Features %s are constant." % constant_features_idx, UserWarning)
C:\Users\Lenovo\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.10_qbz5n2kfra8p0\LocalCache\local-packages\Python310\site-packages\sklearn\feature_selection\_univariate_selection.py:113: RuntimeWarning: invalid value encountered in divide
  f = msb / msw


Eliminação de características recursivas

In [8]:
from sklearn.feature_selection import RFE
from sklearn.svm import SVC

print(X_train.shape)
print(X_test.shape)

# Inicializar o estimador do modelo
estimator = SVC(kernel="linear")

# Inicialize o RFE com o estimador e o número desejado de características a serem mantidas
rfe = RFE(estimator, n_features_to_select=637, step=1)
rfe.fit(X_train, y_train)

# Transforme os conjuntos de treino e teste de acordo com as características selecionadas
X_train_new = rfe.transform(X_train)
X_test_new = rfe.transform(X_test)

print(X_train_new.shape)
print(X_test_new.shape)

(108, 6373)
(48, 6373)
(108, 637)
(48, 637)


Seleção de características usando SelectFromModel

In [9]:
from sklearn.feature_selection import SelectFromModel
from sklearn.svm import SVC

print(X_train.shape)
print(X_test.shape)

# Definindo o modelo que será usado para selecionar características
model = SVC(kernel="linear")

# Inicializando o seletor e ajustando ao conjunto de treino
selector = SelectFromModel(estimator=model, max_features=637)  # Selecionando 99 características
selector.fit(X_train, y_train)

# Aplicando a seleção de características aos conjuntos de treino e teste
X_train_new = selector.transform(X_train)
X_test_new = selector.transform(X_test)

print(X_train_new.shape)
print(X_test_new.shape)

(108, 6373)
(48, 6373)
(108, 637)
(48, 637)


Seleção sequencial de características

In [11]:
from sklearn.feature_selection import SequentialFeatureSelector
from sklearn.svm import SVC

print(X_train.shape)
print(X_test.shape)

# Inicializando o classificador
model = SVC(kernel="linear")

# Inicializando o Sequential Feature Selector
# Defina o número desejado de características e a direção (forward ou backward)
sfs = SequentialFeatureSelector(model, n_features_to_select=637, direction='forward')

# Ajustando o Sequential Feature Selector ao conjunto de treino
sfs.fit(X_train, y_train)

# Obtendo o novo conjunto de características para treino e teste
X_train_new = sfs.transform(X_train)
X_test_new = sfs.transform(X_test)

print(X_train_new.shape)
print(X_test_new.shape)

(108, 6373)
(48, 6373)


Testando Linear Discriminant Analysis (LDA)

In [ ]:
from sklearn.discriminant_analysis import LinearDiscriminantAnalysis
from sklearn.metrics import accuracy_score

# Instanciando o classificador LDA
lda = LinearDiscriminantAnalysis()

# Treinando o modelo LDA com os dados de treinamento
lda.fit(X_train_new, y_train)

# Realizando previsões nos dados de teste
y_pred = lda.predict(X_test_new)

# Calculando a precisão da classificação
accuracy = accuracy_score(y_test, y_pred)
print(f"Acurácia do modelo: {accuracy * 100:.2f}%")

Acurácia do modelo: 64.58%


Testando Decision Tree (DT)

In [ ]:
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import accuracy_score

# Inicializa o classificador de árvore de decisão
clf = DecisionTreeClassifier(random_state=42)

# Treina o classificador com o conjunto de treino
clf.fit(X_train_new, y_train)

# Realiza a previsão com o conjunto de teste
y_pred = clf.predict(X_test_new)

# Calcula a acurácia do modelo
accuracy = accuracy_score(y_test, y_pred)
print(f'Acurácia do modelo: {accuracy * 100:.2f}%')

Acurácia do modelo: 41.67%


Testando Rede Neural Simples (NN)

In [ ]:
from sklearn.neural_network import MLPClassifier
from sklearn.metrics import accuracy_score

# Criando o classificador da rede neural
classifier = MLPClassifier(hidden_layer_sizes=(10,), max_iter=1000, random_state=42)

# Treinando o classificador
classifier.fit(X_train_new, y_train)

# Realizando previsões no conjunto de teste
predictions = classifier.predict(X_test_new)

# Calculando a acurácia do modelo
accuracy = accuracy_score(y_test, predictions)
print(f'Acurácia do modelo: {accuracy * 100:.2f}%')

Acurácia do modelo: 60.42%


Testando Support Vector Machine (SVM)

In [ ]:
from sklearn.svm import SVC
from sklearn.metrics import accuracy_score

# Criar um classificador SVM
classificador_svm = SVC(kernel='linear')  # Você pode escolher o tipo de kernel desejado

# Treinar o classificador SVM
classificador_svm.fit(X_train_new, y_train)

# Fazer previsões usando o conjunto de teste
predictions = classificador_svm.predict(X_test_new)

# Calculando a acurácia do modelo
accuracy = accuracy_score(y_test, predictions)
print(f'Acurácia do modelo: {accuracy * 100:.2f}%')

Acurácia do modelo: 72.92%


Testando Random Forest (RF)

In [ ]:
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score, classification_report

# Criando o classificador Random Forest
classificador_rf = RandomForestClassifier(n_estimators=100, random_state=42)

# Treinando o classificador com os dados de treinamento
classificador_rf.fit(X_train_new, y_train)

# Fazendo previsões com o conjunto de teste
predictions = classificador_rf.predict(X_test_new)

# Calculando a acurácia do modelo
accuracy = accuracy_score(y_test, predictions)
print(f'Acurácia do modelo: {accuracy * 100:.2f}%')

Acurácia do modelo: 56.25%


Resultados obtidos:
|Características                            |LDA   |DT    |NN    |SVM       |RF    |Média     |
|:-----------------------------------------:|:----:|:----:|:----:|:--------:|:----:|:--------:|
|emobase (988)                              |60.42%|41.67%|66.67%|72.92%    |62.50%|60.84%    |
|emobase [Univariate feature selection] (2) |58.33%|64.58%|58.33%|64.58%    |64.58%|__62.08%__|
|emobase [Recursive feature elimination] (2)|45.83%|54.17%|45.83%|45.83%    |54.17%|49.17%    |
|ComParE (6373)                             |60.42%|50.00%|60.42%|64.58%    |60.42%|59.17%    |
|eGeMAPS (88)                               |70.83%|52.08%|62.50%|54.17%    |66.67%|61.25%    |
|Média                                      |62.50%|52.08%|61.98%|__64.06%__|63.54%|59.17%    |